src/Compression/CompressionCodecEncrypted.h

```diff
@@ -1,104 +1,114 @@
+    * blocks in AES-128-GCM-SIV with a universally fixed nonce (12
+    * repeated NUL characters).
+    *
+    * This construct has a weakness due to the nonce being fixed at
+    * all times: when the same data block is encrypted twice, the
+    * resulting ciphertext will be exactly the same. We have to live
+    * with this weakness because ciphertext must be deterministic,
+    * as otherwise our engines like ReplicatedMergeTree cannot
+    * deduplicate data blocks.
+    */
+class CompressionCodecEncrypted : public ICompressionCodec
+{
+public:
+    /** If a key is available, the server is supposed to
+        * invoke this static method at the startup. The codec will
+        * refuse to compress or decompress any data until that. The
+        * key can be an arbitrary octet string, but it is
+        * recommended that the key is at least 16 octets long.
+        *
+        * Note that the key is currently not guarded by a
+        * mutex. This method should be invoked no more than once.
+        */
+    explicit CompressionCodecEncrypted(CompressionMethodByte Algorithm);
+
+    /**
+        * This is utility class. It holds information about encryption configuration.
+        */
+    class Configuration
     {
     public:
-        /** If a master key is available, the server is supposed to
-          * invoke this static method at the startup. The codec will
-          * refuse to compress or decompress any data until that. The
-          * key can be an arbitrary octet string, but it is
-          * recommended that the key is at least 16 octets long.
-          *
-          * Note that the master key is currently not guarded by a
-          * mutex. This method should be invoked no more than once.
-          */
-        static void setMasterKey(const std::string_view & master_key);
-
-        CompressionCodecEncrypted(const std::string_view & cipher);
+        static Configuration & instance();
 
-        uint8_t getMethodByte() const override;
-        void updateHash(SipHash & hash) const override;
+        void load(const Poco::Util::AbstractConfiguration & config, const String & config_prefix);
+        String getKey(UInt64 id) const;
+        String getCurrentKey() const;
+        UInt64 getCurrentKeyID() const;
+        String getNonce() const;
+        CompressionMethodByte getAlgorithmByte() const;
 
-        bool isCompression() const override
-        {
-            return false;
-        }
-
-        bool isGenericCompression() const override
-        {
-            return false;
-        }
-
-        bool isPostProcessing() const override
+    private:
+        struct Params
         {
-            return true;
-        }
-
-    protected:
-        UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
+            std::unordered_map<UInt64, String> keys_storage;
+            UInt64 current_key_id;
+            String nonce;
+            CompressionMethodByte algorithm;
```

Comment:
There can be both `AES_128_GCM_SIV` AND `AES_256_GCM_SIV` existing in the same configuration each with its own keys and nonce and current key ID. Maybe it's a bit strange but I don't see why we could prohibit using both codecs at the same time.